LoopKit/DeviceManager/DeviceLog/PersistentDeviceLog.swift

```diff
@@ -0,0 +1,125 @@
+//
+//  PersistentDeviceLog.swift
+//  LoopKit
+//
+//  Created by Pete Schwamb on 1/13/20.
+//  Copyright © 2020 LoopKit Authors. All rights reserved.
+//
+
+import Foundation
+import CoreData
+import os.log
+
+
+// Using a framework specific class will search the framework's bundle for model files.
+class PersistentContainer: NSPersistentContainer {
+}
+
+public class PersistentDeviceLog {
+    
+    private let storageFile: URL
+    
+    private let managedObjectContext: NSManagedObjectContext
+
+    private let persistentContainer: NSPersistentContainer
+    
+    private let maxAge: TimeInterval
+    
+    private var earliestLogDate: Date {
+        return Date(timeIntervalSinceNow: -maxAge)
+    }
+    
+    private let log = OSLog(category: "PersistentDeviceLog")
+    
+    public init(storageFile: URL, maxAge: TimeInterval = TimeInterval(7 * 24 * 60 * 60)) {
+        self.storageFile = storageFile
+        self.maxAge = maxAge
+
+        managedObjectContext = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+        managedObjectContext.mergePolicy = NSMergeByPropertyStoreTrumpMergePolicy
+        managedObjectContext.automaticallyMergesChangesFromParent = true
+
+        let storeDescription = NSPersistentStoreDescription(url: storageFile)
+        persistentContainer = PersistentContainer(name: "DeviceLog")
+        persistentContainer.persistentStoreDescriptions = [storeDescription]
+        persistentContainer.loadPersistentStores { description, error in
+            if let error = error {
+                fatalError("Unable to load persistent stores: \(error)")
+            }
+        }
+        managedObjectContext.persistentStoreCoordinator = persistentContainer.persistentStoreCoordinator
+    }
+    
+    public func log(deviceManagerIdentifier: String, deviceIdentifier: String?, type: DeviceLogEntryType, message: String, completion: ((Error?) -> Void)? = nil) {
+        managedObjectContext.perform {
+            let entry = DeviceLogEntry(context: self.managedObjectContext)
+            entry.deviceManager = deviceManagerIdentifier
```

Comment:
[nit] Perhaps consider `entry.deviceManagerIdentifier` so that it is consistent with `entry.deviceIdentifier`?